### DISCUSSION TITLE
Wrong energy of TFI with negative sign of an external magnetic field

### DISCUSSION THREAD

**Discussion by abukva** [netket/netket#1383] [General]
Hi, I was trying to compute something and was getting strange results so I've decided to test how RBM deals with simple TFI. Namely the problem is what happens with a ground state energy when you change  the sign of a external magnetic field, from theory we know that this shouldn't have any implications, but in NetKet it does, am I doing something wrong or is this actually a bug?

```
N=3
g = nk.graph.Square(N, pbc=True)

gg = 2.0

hi = nk.hilbert.Spin(0.5)**(g.n_nodes)
ha = nk.operator.LocalOperator(hi)
for i in g.nodes():
    ha -= gg * nk.operator.spin.sigmax(hi,i)
    
for i,j in g.edges():
    ha -= nk.operator.spin.sigmaz(hi,i)*nk.operator.spin.sigmaz(hi,j)

ma = nk.models.RBM( alpha=3)
sa = nk.sampler.MetropolisLocal(hilbert=hi, n_chains_per_rank=1)
op = nk.optimizer.Sgd(learning_rate=0.05)
sr = nk.optimizer.SR(diag_shift=0.1)
vs = nk.vqs.MCState(sa, ma, n_samples_per_rank=1000, n_discard_per_chain=100)

gs = nk.VMC(hamiltonian=ha,
            optimizer=op,
            preconditioner=sr,
            variational_state=vs)
gs.run(out=None, n_iter=100)

```
Then I run this for `gg=2` and `gg=-2`. In the case of `gg=-2` I get `vs.expect(ha)=-1.800e+01 ± nan [σ²=0.000e+00]` and in the case of `gg=2` `vs.expect(ha)=-22.6631 ± 0.0059 [σ²=0.0416]`. In order to verify that the sign of an external magnetic field indeed doesn't make any difference I do a quick sparse exact diagonalization:

```
sp_h=ha.to_sparse()
eig_vals1, eig_vecs1 = eigsh(sp_h, k=2, which="SA")
print("eigenvalues with scipy sparse:", eig_vals1)
```
and in the both cases I get the same result: `eigenvalues with scipy sparse: [-22.68378118 -22.54085076]`

**Comment by gcarleo** [comment]
I converted this to a discussion since it's not a NetKet issue by itself. The reason is quite simple: if you flip the sign of the field you should allow the wave function to take negative signs. If you take an RBM with real weights, this is not the case, the wave function is positive definite.

**Reply by PhilipVinc** [reply]
aka, you need to specify
```python
ma = nk.models.RBM( alpha=3, param_dtype=complex)
```

The optimisation will be harder, however.

**Reply by abukva** [reply]
Ok thanks, I thought this is done by default.

**Reply by abukva** [reply]
I just tried with complex parameters and still have the same result.

**Reply by gcarleo** [reply]
try harder

**Reply by PhilipVinc** [reply]
It's, as often, an initialisation problem.
Put
```python
ma = nk.models.RBM(alpha=3, param_dtype=complex)
...
vs = nk.vqs.MCState(...)
vs.init_parameters(nn.initializers.normal(stddev=0.001))
```

before and it will converge .

![fig](https://user-images.githubusercontent.com/2407108/199736372-2a9364bf-e379-4248-922c-64efa9db3a76.png)

**Reply by abukva** [reply]
Ok, thank you very much, it works now!
